### ISSUE TITLE
Using shuffle or rebatch may cause OOM problem

### ISSUE THREAD

**Issue by liurcme** [alibaba/HybridBackend#43]
# 1. Current behavior

Using shuffle or rebatch may cause OOM problem.

## 1.1 小文件测试记录

total parquet file count: 15780

total parquet file size: 126G

total sample count: 600w

No. | Test Scenarios | RAM Usage
-- | -- | --
1 | no shuffle  no rebatch | use 50G ram stable
2 | shuffle，buffer_size=2048  rebatch，batch_size=8192 | start at 53G ram,  and increasingly  use more ram rapidly, exceed 94G ram limit within 2 minutes
3 | shuffle，buffer_size=8  rebatch，batch_size=8192 | increasingly  use more ram slowly, exceed 94G ram limit after 2 hours
4 | no shuffle  rebatch，batch_size=8192 | increasingly  use more ram slowly, exceed 94G ram limit after 2 hours
5 | shuffle，buffer_size=8 no rebatch | use 50G ram stable at first, but increasingly use more ram after 1 hour

![image-20220303211253747](https://user-images.githubusercontent.com/100847992/156573483-5aee4e07-0be0-4808-8f48-d8ae20f41929.png)

![image-20220303211236154](https://user-images.githubusercontent.com/100847992/156573511-e6e90b2e-1de6-49d1-83f9-c024d87d508c.png)

![image-20220303211125580](https://user-images.githubusercontent.com/100847992/156573541-6116b3db-0f8d-4591-9154-5f2107902211.png)

![image-20220303211314882](https://user-images.githubusercontent.com/100847992/156573562-e7b92233-9067-43e7-830f-67c366b89c6f.png)

![image-20220303211215059](https://user-images.githubusercontent.com/100847992/156573584-b08154ea-0369-4843-b79b-eacb8d069fcb.png)

## 1.2 大文件测试记录
total parquet file count: 240

total parquet file size: 126G

单个parquet文件大小：500MB

total sample count: 600w

![微信图片_20220323124551](https://user-images.githubusercontent.com/100847992/159626835-3711ae5f-bda4-4a16-a443-e84d4cc5ab61.jpg)

![微信图片_20220323124557](https://user-images.githubusercontent.com/100847992/159626843-5be54d4e-c328-4cd3-81cb-e5a3da3d54b7.jpg)

现象：每个epoch训练结束时，有明显内存回收的过程，但是回收不干净，导致每个epoch后使用的内存峰值越来越多，最终OOM。但是如果1、2个epoch内能训练完，内存不会爆。

## 1.3 不同训练方式测试对比

|              | 新sdk         | 旧sdk                                                        |
| ------------ | ------------- | ------------------------------------------------------------ |
| 是否内存溢出 | 是            | 否                                                           |
| 环境         | T4 单机单卡   | T4 单机单卡                                                  |
| 训练方式     | session.run() | tf.estimator.train_and_evaluate(estimator, train_spec,  eval_spec) |
| 调用链       | 见下图1       | 见下图2                                                      |

![微信图片_20220323124545](https://user-images.githubusercontent.com/100847992/159626803-9120171f-a170-44cc-a3f8-25682486120f.png)

![微信图片_20220323124542](https://user-images.githubusercontent.com/100847992/159626816-21f6ac6f-8885-448c-a0a5-0ac9ac22a4d9.png)

# 2.  Expected behavior

During the training process, tensorflow should use stable amount of RAM, not using more and more RAM.

# 3.  System information
-   **OS Platform and Distribution**: Ubuntu 18.04.5 LTS
-   **TensorFlow version**: 1.15.0
-   **Python version**: 3.6
-   **CUDA/cuDNN version**: 10.1
-   **RAM**: 94G
-   **GPU model and memory**: Tesla T4, 16G

# 4. Code to reproduce

```python
BATCH_SIZE = 8192

parquet_file_list = ['some_parquet_file1.snappy.parquet', 'some_parquet_file2.snappy.parquet', ...]
filenames_ds = tf.data.TFRecordDataset.from_tensor_slices(file_list)
hb_fields = []
hb_fields.append(hb.data.DataFrame.Field('int_field', tf.int64, ragged_rank=0))
hb_fields.append(hb.data.DataFrame.Field('float_field', tf.float32, ragged_rank=0))
hb_fields.append(hb.data.DataFrame.Field('array_field', tf.float32, ragged_rank=1))   # ... and some anthor fields


# 1. no shuffle, no rebatch
ds = filenames_ds.apply(hb.data.read_parquet(BATCH_SIZE, hb_fields, num_parallel_reads=tf.data.experimental.AUTOTUNE))

# 2. big shuffle and rebatch
ds = filenames_ds.apply(hb.data.read_parquet(BATCH_SIZE, hb_fields, num_parallel_reads=tf.data.experimental.AUTOTUNE))
ds.shuffle(2048)
ds.apply(hb.data.rebatch(BATCH_SIZE, fields=fields))

# 3. small shuffle and rebatch
ds = filenames_ds.apply(hb.data.read_parquet(BATCH_SIZE, hb_fields, num_parallel_reads=tf.data.experimental.AUTOTUNE))
ds.shuffle(8)
ds.apply(hb.data.rebatch(BATCH_SIZE, fields=fields))

# 4. no shuffle and rebatch
ds = filenames_ds.apply(hb.data.read_parquet(BATCH_SIZE, hb_fields, num_parallel_reads=tf.data.experimental.AUTOTUNE))
ds.apply(hb.data.rebatch(BATCH_SIZE, fields=fields))

# 5. small shuffle and no rebatch
ds = filenames_ds.apply(hb.data.read_parquet(BATCH_SIZE, hb_fields, num_parallel_reads=tf.data.experimental.AUTOTUNE))
ds.shuffle(8)
```

# Willing to contribute

Yes

**Comment by 2sin18** [comment]
Thanks for your contribution, I would try to reproduce and fix it.

**Comment by 2sin18** [comment]
I cannot reproduce this issue. More details might be helpful.

I use modified dataset_benchmark to reproduce multiple epoch data access:

```bash
python -m hybridbackend.tools.dataset_benchmark ./train.parquet --epochs=100 --num-steps=1000000
```

I use pidstat to monitor RSS:

```bash
pidstat -r -p ${PID} 1 | tee ${PID}.pidstat.log
```

I tried 3 cases as below:

Case 1: no rebatch + [criteo dataset](https://www.kaggle.com/mrkmakr/criteo-dataset)

![image](https://user-images.githubusercontent.com/320794/162382133-d0924184-f489-425c-ab61-1bc6a6b14808.png)

Case 2: rebatch + [criteo dataset](https://www.kaggle.com/mrkmakr/criteo-dataset)

(Add `ds = ds.apply(hb.data.rebatch(params.batch_size * 2))` to benchmark code.)

![image](https://user-images.githubusercontent.com/320794/162382285-a45827b8-d0cc-4694-b553-808f4b25463b.png)

Case 3: rebatch + [ele.me online search dataset](https://tianchi.aliyun.com/dataset/dataDetail?dataId=120281) (has sequence data type)

![image](https://user-images.githubusercontent.com/320794/162382552-892b938d-3063-44f9-bd07-de4e1d013661.png)

**Comment by liurcme** [comment]
### 1. Current behavior
batch_size matters.
small batch_size: no memory leak or slight leak, which is acceptable
large batch_size: severe memory leak

### 2. Code to reproduce
System information see above.

(1) first install tensorflow and hybridbackend
```shell
pip3 install tensorflow==1.15.0
pip3 install hybridbackend==0.5.2
```

(2) run following code, notice batch_size and rebatch_size are the same.
```python
import tensorflow as tf
import tensorflow.keras as keras
import hybridbackend.tensorflow as hb

BATCH_SIZE = 8192

def get_parquet_ds():
    filenames_ds = tf.data.Dataset.from_tensor_slices([
        'training_sample_file.snappy.parquet1',
        'training_sample_file.snappy.parquet2'
    ])
    hb_fields = []

    def _map(elem):
        features = {
            "int_feature": tf.cast(tf.reshape(elem["int_feature"], [-1, 1]), dtype=tf.float32),
            "array_feature": tf.cast(tf.reshape(elem["array_feature"].values, [-1, 50]),
                                              dtype=tf.float32)
        }
        labels = tf.reshape(elem["label"], [-1, 1])
        return features, labels

    hb_fields.append(hb.data.DataFrame.Field("int_feature", tf.int64, ragged_rank=0))
    hb_fields.append(hb.data.DataFrame.Field("array_feature", tf.int64, ragged_rank=1))
    hb_fields.append(hb.data.DataFrame.Field("label", tf.float32, ragged_rank=0))
    iterator = filenames_ds.apply(
        hb.data.read_parquet(BATCH_SIZE, hb_fields, num_parallel_reads=tf.data.experimental.AUTOTUNE))
    iterator = iterator.apply(hb.data.rebatch(BATCH_SIZE, fields=hb_fields)).map(_map).repeat(10)

    return iterator


def train():
    feature1 = tf.compat.v1.placeholder(tf.float32, shape=[None, 1])
    feature2 = tf.compat.v1.placeholder(tf.float32, shape=[None, 50])
    labels = tf.compat.v1.placeholder(tf.float32, shape=[None, 1])
    inputs = tf.concat([feature1, feature2], axis=1)

    with tf.name_scope(name="dense"):
        dense = keras.layers.Dense(128, activation="relu")(inputs)
        dense = keras.layers.Dense(56, activation="relu")(dense)

    with tf.name_scope(name="output"):
        output = keras.layers.Dense(1, activation="softmax")(dense)

    with tf.name_scope(name="loss"):
        loss_mse = tf.compat.v1.losses.mean_squared_error(
            labels=labels,
            predictions=output)

    adam = tf.compat.v1.train.AdamOptimizer(learning_rate=0.01)
    optimize_op = adam.minimize(loss_mse)
    global_init_op = tf.compat.v1.global_variables_initializer()

    ds = get_parquet_ds()
    iterator = ds.make_one_shot_iterator()
    get_data_op = iterator.get_next()

    with tf.Session() as sess:
        sess.run([global_init_op])
        i = 1
        while True:
            try:
                sample = sess.run([get_data_op])

                f_category = sample[0][0]["int_feature"]
                f_list = sample[0][0]["array_feature"]
                labels_ = sample[0][1]
                train_loss, _ = sess.run([loss_mse, optimize_op], feed_dict={
                    feature1: f_category,
                    feature2: f_list,
                    labels: labels_
                })

                if i % 1000 == 0:
                    print(f"step={i}, train_loss={train_loss}")
                i += 1
            except tf.errors.OutOfRangeError:
                break


if __name__ == '__main__':
    train()

```

**Comment by 2sin18** [comment]
Thanks but I still cannot repeat your problem using above code:

```bash
05:29:20 PM   UID       PID  minflt/s  majflt/s     VSZ    RSS   %MEM  Command
05:29:21 PM 66732    125250 417836.00      0.00 52524328 5847180   1.11  python
05:40:26 PM 66732    125250 374475.00      0.00 55452364 5561484   1.05  python
05:49:55 PM 66732    125250 486975.00      0.00 56272076 5438272   1.03  python
05:55:00 PM 66732    125250 443005.00      0.00 56739020 5270048   1.00  pytho
```

Test data is generated from below code:

```python
import numpy as np
import pandas as pd
import random

data_list = []
for i in range(1, 1000000):
    int_feature = random.randint(1, 1000)
    array_feature = [random.randint(1, 1000) for x in range(0, 50)]
    data_list.append([int_feature, array_feature, 0.8])

df = pd.DataFrame(data_list, columns=["int_feature", "array_feature", "label"])
df['label'] = pd.to_numeric(df["label"], downcast="float")
df.to_parquet("parquet_sample_file.parquet")
```

**Comment by 2sin18** [comment]
The problem cannot be reproduced for data on local disk, it seems that Java GC configuration might be related.
